pep-0697.rst

```diff
@@ -0,0 +1,357 @@
+a heap type (``PyType_Spec``) the example reads:
+
+.. code-block:: c
+
+    typedef struct {
+        PyListObject list;
+        int state;
+    } SubListObject;
+
+    static PyType_Spec Sublist_spec = {
+        .name = "sublist.SubList",
+        .basicsize = sizeof(SubListObject),
+        .itemsize = 0,
+        .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+        .slots = SubList_slots
+    };
+
+Since the superclass struct (``PyListObject``) is part of the subclass struct
+(``SubListObject``):
+
+- ``PyListObject`` size must be known at compile time, and
+- the size must be the same across all interpreters/versions the compiled
+  extension is ABI-compatible with.
+
+But in limited API/stable ABI, we do not expose the size of ``PyListObject``,
+so that it can vary between CPython versions (and even between possible
+alternate ABI-compatible C API implementations).
+
+With the size not available, limited API users must resort to workarounds such
+as querying ``__basicsize__`` and plugging it into ``PyType_Spec`` at runtime,
+and divining the correct offset for their extra data.
+This requires making assumptions about memory layout, which the limited API
+is supposed to hide.
+
+
+Extending variable-size objects
+-------------------------------
+
+Another scenario where the traditional way to extend an object does not work
+is variable-sized objects, i.e. ones with non-zero ``tp_itemsize``.
+If the instance struct ends with a variable-length array (such as
+in ``tuple`` or ``int``), subclasses cannot add their own extra data without
+detailed knowledge about how the superclass allocates and uses its memory.
+
+Some types, such as CPython's ``PyHeapType``, handle this by storing
+variable-sized data after the fixed-size struct.
+This means that any subclass can add its own fixed-size data.
+(Only one class in the inheritance hierarchy can use variable-sized data, though.)
+This PEP proposes API that makes this practice easier, and ensures the
+variable-sized data is properly aligned.
+
+Note that many variable-size types, like ``int`` or ``tuple``, do not use
+this mechanism. 
+This PEP does not propose any changes to existing variable-size types (like
+``int`` or ``tuple``) except ``PyHeapType``.
+
+
+Extending ``PyHeapType`` specifically
+-------------------------------------
+
+The motivating problem this PEP solves is creating metaclasses, that is,
+subclasses of ``type``.
+The underlying ``PyHeapTypeObject`` struct is both and variable-sized, and
+opaque in the limited API.
+
+Projects such as language bindings and frameworks that need to attach custom
+data to metaclasses currently resort to questionable workarounds.
+The situation is worse in projects that target the Limited API.
+
+For an example of the currently necessary workarounds, see:
+`nb_type_data_static <https://github.com/wjakob/nanobind/blob/f3044cf44763e105428e4e0cf8f42d951b9cc997/src/nb_type.cpp#L1085>`_
```

Comment:
Can't say I understand what's going on there, since it's modern C++ and I don't read that.

Could you perhaps explain what it is doing?

Are you aware of pure-C examples of similar workarounds? Or is C++ the main target here?

Reply:
It's querying the Python attribute ``__basicsize__`` and plugging it into ``PyType_Spec`` used to create a subclass. I'll move the English description here.

I'm using nanobind (C++) for examples & initial design discussions, because it's small and free of historical baggage. But Java or Rust bindings have similar use cases, and I'll consider the PEP a failure if JPype & Pyo3 don't find it useful.

(My own motivation is improving subclassing API in general, but that's not what people behind binding generators are asking for...)

Reply:
Just don't expect your readers to be familiar with nanobind. It's fine to explain that nanobind provides a use case, but you still need to explain what problem it is solving and how, and how its solution is sub-optimial and how your PEP provides a better solution. I'm guessing from what you explained in comments above that when a subclass of a variable-sized type (like tuple) needs to add extra state it must define a struct to hold the extra state and write a custom function that calculates the offset of that extra state relative to the nominal (`PyObject *`) start of the object, and it also needs to override the new-instance allocation (`tp_new`? `tp_alloc`?) to allocate space for the extra state.

Reply:
The C equivalent snippet of the code would be helpful. The code is small and easy is to write in C.